File: packages/python/src/randino/decorate/_attach.py
"""Shared by `rand_suffix` and `rand_prefix`."""

import math
from collections.abc import Callable

from randino._internal.utils import clamp, rand_token
from randino.decorate.data import AFFIX_CHARSET, AFFIX_LENGTH_MAX


def attach(
    value: str | list[str] | None,
    length: int,
    separator: str,
    charset: str,
    join: Callable[[str, str, str], str],
) -> str | list[str]:
    """Attach a freshly drawn token to one string, or to every string in a list.

    One token each rather than one for the batch, which is the whole point of
    passing a generator's output here. `join` is what decides the side it lands
    on. With no value at all the bare token is the answer, separator and all left
    off — what a decorator attaches is worth having on its own.

    Args:
        value: The string, the list of strings, or None for the token alone.
        length: Characters in the token, clamped to `1..AFFIX_LENGTH_MAX`.
        separator: Placed between the value and the token.
        charset: Characters the token is drawn from; empty means the default.
        join: Given `(value, token, separator)`, returns the finished string.

    Returns:
        A string when `value` is a string or None, a list when it is a list.
    """
    size = clamp(math.floor(length), 1, AFFIX_LENGTH_MAX)
    alphabet = charset or AFFIX_CHARSET

    def token() -> str:
        return rand_token(size, alphabet)

    if value is None:
        return token()

    def one(item: str) -> str:
        return join(item, token(), separator)

    if isinstance(value, str):
        return one(value)

    return [one(item) for item in value]

File: packages/javascript/lib/decorate/attach.ts
// Shared by `randSuffix` and `randPrefix`. Internal — the two of them differ by
// one line, which is the side the token lands on.

import { resolveWhole } from '../_internal/generate.js';
import { randToken } from '../_internal/utils.js';
import type { RandAffixOptions } from '../_types/global.js';
import {
	AFFIX_CHARSET,
	AFFIX_LENGTH_DEFAULT,
	AFFIX_LENGTH_MAX,
	AFFIX_SEPARATOR_DEFAULT
} from './data/index.js';

/**
 * Tell a value from an options object in the first argument. Every decorator
 * takes both shapes: what it attaches is worth having on its own, so the value
 * is optional and `randSuffix({ length: 8 })` has to mean the options.
 */
export function firstArgument<T>(
	value: string | string[] | T | undefined,
	options: T | undefined,
	fallback: T
): { target: string | string[] | undefined; settings: T } {
	const isValue = typeof value === 'string' || Array.isArray(value);

	return {
		target: isValue ? (value as string | string[]) : undefined,
		settings: (isValue ? options : (value as T | undefined)) ?? fallback
	};
}

/**
 * Attaches a **freshly drawn** token to one string, or to every string in an
 * array — one token each rather than one for the batch, which is the whole
 * point of passing a generator's output here. With no value at all it hands
 * back the bare token, separator and all left off.
 */
export function attach(
	value: string | string[] | RandAffixOptions | undefined,
	options: RandAffixOptions | undefined,
	join: (value: string, token: string, separator: string) => string
): string | string[] {
	const { target, settings } = firstArgument<RandAffixOptions>(value, options, {});
	// A `length` that is not a whole number is the default rather than no token at
	// all: `NaN` clamps to `NaN`, and a loop that runs `NaN` times writes nothing,
	// so `randSuffix('x', { length: NaN })` used to hand back `'x_'`.
	const length = resolveWhole(settings.length, AFFIX_LENGTH_DEFAULT, 1, AFFIX_LENGTH_MAX);
	const charset =
		typeof settings.charset === 'string' && settings.charset ? settings.charset : AFFIX_CHARSET;
	const separator =
		typeof settings.separator === 'string' ? settings.separator : AFFIX_SEPARATOR_DEFAULT;
	const token = () => randToken(length, charset);

	if (target === undefined) {
		return token();
	}

	const one = (item: string) => join(item, token(), separator);

	return Array.isArray(target) ? target.map(one) : one(target);
}
